File: Magick++/demo/piddle.cpp
// This may look like C code, but it is really -*- C++ -*-
//
// Copyright Bob Friesenhahn, 1999, 2000
//
// PerlMagick "piddle" demo re-implemented using Magick++ methods.
// The PerlMagick "piddle" demo is written by John Cristy
//

#include <string>
#include <iostream>

#include <Magick++.h>

using namespace std;

using namespace Magick;

int main( int /*argc*/, char ** argv)
{

  // Initialize ImageMagick install location for Windows
  MagickIncarnate(*argv);

  string srcdir("");
  if(getenv("srcdir") != (char*)NULL)
    srcdir = getenv("srcdir") + string("/");

  // Common font to use.
  string font = string("@") + srcdir + "Generic.ttf";

  try {

    //
    // Create a 300x300 white canvas.
    //
    Image image( "300x300", "white" );

    //
    // Draw blue grid
    //
    {
      std::list<Magick::Drawable> drawList;
      drawList.push_back(DrawableStrokeColor("#ccf"));
      for ( int i=0; i < 300; i += 10 )
        {
          drawList.push_back(DrawableLine(i,0, i,300));
          drawList.push_back(DrawableLine(0,i, 300,i));
        }
      image.draw(drawList);
    }

    //
    // Draw rounded rectangle.
    //
    {
//       image.fillColor("blue");
//       image.strokeColor("red");
//       image.draw(DrawableRoundRectangle(15,15, 70,70, 10,10));

//       image.fillColor("blue");
//       image.strokeColor("maroon");
//       image.lineWidth(4);
//       image.draw(DrawableRoundRectangle(15,15, 70,70, 10,10));

      std::list<Magick::Drawable> drawList;
      drawList.push_back(DrawableFillColor("blue"));
      drawList.push_back(DrawableStrokeColor("red"));
      drawList.push_back(DrawableRoundRectangle(15,15, 70,70, 10,10));

      drawList.push_back(DrawableFillColor("blue"));
      drawList.push_back(DrawableStrokeColor("maroon"));
      drawList.push_back(DrawableStrokeWidth(4));
      drawList.push_back(DrawableRoundRectangle(15,15, 70,70, 10,10));
      image.draw(drawList);
    }

    //
    // Draw curve.
    //
    {
      std::list<Magick::Coordinate> points;
      image.strokeColor("black");
      image.fillColor(Color()); // Unset
      image.lineWidth(4);
      points.push_back(Coordinate(20,20));
      points.push_back(Coordinate(100,50));
      points.push_back(Coordinate(50,100));
      points.push_back(Coordinate(160,160));
      
      image.draw(DrawableBezier(points));
    }

    //     cout << "Writing image \"piddle.miff\" ..." << endl;
    //     image.write( "piddle.miff" );

    cout << "Display image..." << endl;
    image.display( );

  }
  catch( Exception error_ )
    {
      cout << "Caught exception: " << error_.what() << endl;
      return 1;
    }
  catch( exception error_ )
    {
      cout << "Caught exception: " << error_.what() << endl;
      return 1;
    }
  
  return 0;
}

File: PerlMagick/demo/piddle.pl
#!/usr/local/bin/perl
# Piddle example using PerlMagick methods.

use Image::Magick;

#
# Create white canvas.
#
$image=Image::Magick->new(size=>'300x300');
$image->Read('xc:white');
#
# Draw blue grid
#
for ($i=0; $i < 300; $i+=10)
{
  $image->Draw(primitive=>'line',points=>"$i,0 $i,300",stroke=>"#ccf");
  $image->Draw(primitive=>'line',points=>"0,$i 300,$i",stroke=>"#ccf");
}
#
# Draw rounded rectangle.
#
$image->Draw(primitve=>'RoundRectangle',fill=>blue,stroke=>'red',
  points=>'65,65 70,70 10,10');
$image->Draw(primitve=>'roundRectangle',stroke=>maroon,linewidth=>4,
  points=>'65,65 70,70 10,10');
#
# Draw curve.
#
$image->Draw(primitve=>'bezier',points=>'20,20, 100,50, 50,100, 160,160',
  stroke=>black,linewidth=>4);
#
# Draw line.
#
$image->Draw(primitive=>'line',points=>"10,200 20,190",stroke=>red);
#
# Draw arc within a circle.
#
$image->Draw(primitive=>'circle',stroke=>'black',fill=>'yellow',linewidth=>4,
  points=>"160,70 200,70");
$image->Draw(primitve=>'Arc',stroke=>'black',fill=>'blue',linewidth=>4,
  points=>'120,30 200,100 45,90');
#
# Draw pentogram.
#
$image->Draw(primitive=>'polygon',
  points=>"160,120 130,190 210,145 110,145 190,190 160,120",stroke=>red,
  fill=>LimeGreen,linewidth=>3);
#
# Draw rectangle.
#
$image->Draw(primitve=>'rectangle',points=>'200,200 260,260',stroke=>yellow,
  linewidth=>5);
$image->Draw(primitve=>'line',points=>'200,260 260,260',stroke=>green,
  linewidth=>5);
$image->Draw(primitve=>'line',points=>'260,200 260,260',stroke=>red,linewidth=>5);
#
# Draw text.
#
$image->Annotate(text=>'This is a test!',geometry=>'+30+140',
  font=>'@Generic.ttf',fill=>'green',pointsize=>24,rotate=>45.0);
$image->Display();
